### PR TITLE
[fix] Minting issues (1593 and 1405)

### LINKED ISSUE #1405: Support minting of assets with AssetValueType::BigQuantity
Precondition:
Registered asset with type AssetValueType::BigQuantity

Expected:
When instruction to mint sent it is committed and asset minted

Current:
When instruction to mint sent then peer reject transaction and  error can be found in logs
```
Transaction 6b6066a7729de5c36cfb3956bbf46f3f6b3a439d7cf4623db424544fcfbd8411 was rejected by reason: `Failed: `Failed to assert type. Caused by: Failed to assert type of asset. Caused by: Asset type error: expected asset of type Quantity, but got BigQuantity` during execution of instruction: jp.co.soramitsu.iroha2.generated.datamodel.isi.Instruction.Mint
```

### LINKED ISSUE #1593: Assets with asset value type `Fixed` can be minted and burned with negative or zero amount
Precondition: some asset with `AssetValueType` is `Fixed` has registered and client sends transaction with instruction to mint with negative amount

Expected: some error returned from the peer

Actual: instruction applied and balance can be negative

### PR DESCRIPTION
**Pull request by appetrosyan**
Signed-off-by: Aleksandr <a-p-petrosyan@yandex.ru>

### Description of the Change

Made every operation involving `iroha_data_model::fixed::Fixed` fail if any value involved in said operation corresponds to a negative `Value`. 

### Issue

Closes #1593 
Closes #1405 

### Benefits

- The checked operations will cause an error if a negative value is passed. 
- The maximum error guarantee is verified using an `ignored` test. (this is only necessary once). 
- Some error messages have been improved. 

### Possible Drawbacks

None

### Alternate Designs 

Check if `Fixed` values are negative only when performing the `Mint` and `Burn` operations, without checking for negative values anywhere within the code.

### CODE REVIEW

File: data_model/src/fixed.rs [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+
+impl Ord for Fixed {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.0.cmp(&other.0)
+    }
+}
+
+impl Encode for Fixed {
+    fn size_hint(&self) -> usize {
+        std::mem::size_of::<Base>()
```
Comment (eadventurous) [inline data_model/src/fixed.rs:140]: Maybe import `std::mem`? It is used in many places.

File: data_model/src/fixed.rs [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+    #[inline]
+    /// Return the only possible negative [`Fixed`] value. Only used for tests.
+    ///
+    /// # Panics
+    /// Never.
+    pub fn negative_one() -> Self {
+        #[allow(clippy::unwrap_used)]
+        Self("-1".parse().unwrap())
+    }
```
Comment (eadventurous) [inline data_model/src/fixed.rs:48]: Should we add `cfg[test]`?

**Review by eadventurous — DISMISSED** [review]

**Review by eadventurous — APPROVED** [review]

File: data_model/src/fixed.rs [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+/// Custom error type for Fixed point operation errors.
+#[derive(Debug, Clone, Error)]
+pub enum FixedPointOperationError {
+    /// All [`Fixed`] values should be positive.
+    #[error("Negative value {0}")]
+    NegativeValue(FixNum),
+    /// Conversion failed.
+    #[error("Failed to produce fixed point number")]
+    FixNum(#[source] ConvertError),
```
Comment (mversic) [inline data_model/src/fixed.rs:99]: could we name this variant more in a way it reflects it's meaning? like `Conversion` or something similar

File: data_model/src/fixed.rs [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+        }
+    }
+
+    #[inline]
+    /// Checked addition
+    ///
+    /// # Errors
+    /// If either of the operands is negative or if addition overflows.
+    pub fn checked_add(self, rhs: Self) -> Result<Self, FixedPointOperationError> {
+        match self.valid()?.0.cadd(rhs.valid()?.0) {
```
Comment (mversic) [inline data_model/src/fixed.rs:72]: do we have to check both operands are valid before the operation? can't the `Fixed` struct be assumed to be in the valid state?

Comment (mversic) [reply data_model/src/fixed.rs:72]: for `checked_sub` as well

File: data_model/src/fixed.rs [hunk 5]
```diff
@@ -0,0 +1,265 @@
+//! Types used for Fixed-point operations. Uses [`fixnum::FixedPoint`].
+#![allow(variant_size_differences)]
```
Comment (mversic) [inline data_model/src/fixed.rs:2]: would you place this over the `FixedPointOperationError` enum if it's required there?
Is there really such a difference in variant sizes for `FixedPointOperationError`?

File: data_model/src/fixed.rs (hunk 4 above)
Comment (appetrosyan) [reply data_model/src/fixed.rs:72]: Not at this point. It's impossible to guarantee that no value will ever be created, using e.g. `unsafe`. That's why both values must be validated before any checked operation.

File: data_model/src/fixed.rs (hunk 5 above)
Comment (appetrosyan) [reply data_model/src/fixed.rs:2]: > place... 
yes. 
> really such difference

23 times. I could use a `Box` `fixnum::ConversionError`.

**Review by mversic — APPROVED** [review]

**Review by eadventurous — APPROVED** [review]

**Pull request merged**